### PR TITLE
WIP: Pubsub pull operator

### PR DESCRIPTION
**Pull request by ANiteckiP**
TODO before merging:
- [ ] Design and implement system tests for the new PubSubPullOperator

---
Issue link: WILL BE INSERTED BY [boring-cyborg](https://github.com/kaxil/boring-cyborg)

Make sure to mark the boxes below before creating PR: [x]

- [ ] Description above provides context of the change
- [ ] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [ ] Unit tests coverage for changes (not needed for documentation changes)
- [ ] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [ ] Relevant documentation is updated including usage instructions.
- [ ] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: airflow/providers/google/cloud/hooks/pubsub.py [hunk 1]
```diff
@@ -529,9 +530,12 @@ def acknowledge(
         :param subscription: the Pub/Sub subscription name to delete; do not
             include the 'projects/{project}/topics/' prefix.
         :type subscription: str
-        :param ack_ids: List of ReceivedMessage ackIds from a previous pull
-            response
+        :param ack_ids: List of ReceivedMessage ackIds from a previous pull response.
+            Mutually exclusive with `messages` argument.
```
Comment (mik-laj) [inline airflow/providers/google/cloud/hooks/pubsub.py]:
```suggestion
            Mutually exclusive with ``messages`` argument.
```

File: airflow/providers/google/cloud/hooks/pubsub.py [hunk 2]
```diff
@@ -529,9 +530,12 @@ def acknowledge(
...
         :type ack_ids: list
+        :param messages: List of ReceivedMessage objects to acknowledge.
+            Mutually exclusive with `messages` argument.
```
Comment (mik-laj) [inline airflow/providers/google/cloud/hooks/pubsub.py]:
```suggestion
            Mutually exclusive with ``messages`` argument.
```

File: airflow/providers/google/cloud/operators/pubsub.py [hunk 3]
```diff
@@ -674,3 +675,130 @@ def execute(self, context):
...
+        If you need to wait for messages, please use
+        :class:`airflow.providers.google.cloud.sensors.PubSubPullSensor`
+        instead.
+
+    .. seealso::
+        For more information on how to use this operator, take a look at the guide:
+        :ref:`howto/operator:PubSubPullSensor`
+
+    .. seealso::
+        :class:`airflow.providers.google.cloud.sensors.PubSubPullSensor`
```
Comment (mik-laj) [inline airflow/providers/google/cloud/operators/pubsub.py]: Can you add short sentences before link?

File: airflow/providers/google/cloud/operators/pubsub.py [hunk 4]
```diff
@@ -674,3 +675,130 @@ def execute(self, context):
         self.log.info("Publishing to topic %s", self.topic)
         hook.publish(project_id=self.project_id, topic=self.topic, messages=self.messages)
         self.log.info("Published to topic %s", self.topic)
+
+
+class PubSubPullOperator(BaseOperator):
+    """Pulls messages from a PubSub subscription and passes them through XCom.
+
+    This Operator always calls PubSub API with return_immediately=True.
+        This means that the graph execution will continue regardless of
```
Comment (mik-laj) [inline airflow/providers/google/cloud/operators/pubsub.py]:
```suggestion
    This means that the graph execution will continue regardless of
```

Comment (mik-laj) [reply airflow/providers/google/cloud/operators/pubsub.py]: This is not necessary.

File: airflow/providers/google/cloud/operators/pubsub.py [hunk 5]
```diff
@@ -674,3 +675,130 @@ def execute(self, context):
         self.log.info("Publishing to topic %s", self.topic)
         hook.publish(project_id=self.project_id, topic=self.topic, messages=self.messages)
         self.log.info("Published to topic %s", self.topic)
+
+
+class PubSubPullOperator(BaseOperator):
+    """Pulls messages from a PubSub subscription and passes them through XCom.
+
+    This Operator always calls PubSub API with return_immediately=True.
```
Comment (mik-laj) [inline airflow/providers/google/cloud/operators/pubsub.py]:
```suggestion
    This Operator always calls PubSub API with ``return_immediately`` equal ``True``.
```

File: airflow/providers/google/cloud/operators/pubsub.py [hunk 6]
```diff
@@ -674,3 +675,130 @@ def execute(self, context):
...
+    @apply_defaults
+    def __init__(
+            self,
+            project_id: str,
+            subscription: str,
+            max_messages: int = 5,
+            ack_messages: bool = False,
+            gcp_conn_id: str = 'google_cloud_default',
+            delegate_to: Optional[str] = None,
+            messages_callback: Optional[Callable[[List[ReceivedMessage], Dict[str, Any]], Any]] = None,
```
Comment (mik-laj) [inline airflow/providers/google/cloud/operators/pubsub.py]: Can you keep gcp_conn_id and delegate_to as last parameters?

File: airflow/providers/google/cloud/sensors/pubsub.py [hunk 7]
```diff
@@ -35,6 +36,9 @@ class PubSubPullSensor(BaseSensorOperator):
         For more information on how to use this operator, take a look at the guide:
         :ref:`howto/operator:PubSubPullSensor`
 
+    .. seealso::
+        :class:`airflow.providers.google.cloud.operators.PubSubPullOperator`
```
Comment (mik-laj) [inline airflow/providers/google/cloud/sensors/pubsub.py]: Can you add some more info before link?

File: airflow/providers/google/cloud/hooks/pubsub.py [hunk 8]
```diff
@@ -545,8 +549,22 @@ def acknowledge(
...
+
         if not project_id:
             raise ValueError("Project ID should be set.")
+
+        if ack_ids is not None and messages is not None:
+            raise ValueError("'ack_ids' and 'messages' arguments are mutually exclusive.")
+
+        if ack_ids is None:
+            if messages is None:
+                raise ValueError("Either 'ack_ids' or 'messages' argument have to be provided.")
```
Comment (nuclearpinguin) [inline airflow/providers/google/cloud/hooks/pubsub.py]: I think this case is handled in 556. At this point we are sure that `ack_ids` or `messages` is not None/

Comment (ANiteckiP) [reply airflow/providers/google/cloud/hooks/pubsub.py]: At 556, we assure that both are not provided at the same time.
Here we check that at least one of them is provided.
Together these two checks assure that one and only one of them is provided.

File: airflow/providers/google/cloud/sensors/pubsub.py (hunk 7 above)
Comment (ANiteckiP) [reply airflow/providers/google/cloud/sensors/pubsub.py]: Added

File: airflow/providers/google/cloud/operators/pubsub.py (hunk 6 above)
Comment (ANiteckiP) [reply airflow/providers/google/cloud/operators/pubsub.py]: Fixed

File: airflow/providers/google/cloud/operators/pubsub.py (hunk 3 above)
Comment (ANiteckiP) [reply airflow/providers/google/cloud/operators/pubsub.py]: Added

File: airflow/providers/google/cloud/hooks/pubsub.py (hunk 8 above)
Comment (ANiteckiP) [reply airflow/providers/google/cloud/hooks/pubsub.py]: An equivalent if-tree would be:

```py
        if ack_ids is not None and messages is None:
            pass
        elif ack_ids is None and messages is not None:
            ack_ids = [
                message.ack_id
                for message in messages
            ]
        else:
            raise ValueError("One and only one of 'ack_ids' and 'messages' arguments have to be provided")
```

If you consider it more readable than current one, I can change it.

Comment (nuclearpinguin) [reply airflow/providers/google/cloud/hooks/pubsub.py]: +1 for the if-tree

Comment (ANiteckiP) [reply airflow/providers/google/cloud/hooks/pubsub.py]: Done

**Comment by ANiteckiP** [comment]
Opened PR against Apache master.
https://github.com/apache/airflow/pull/7613

**Pull request closed without merging**